### DISCUSSION TITLE
Virtual joystick for TouchImputDevice

### DISCUSSION THREAD

**Discussion by llama-nl** [godotengine/godot-proposals#10001] [2D]
I was using an addon called Vistaul Joystick to achieve joystick controller. But I am not feeling too comfortable to use an addon to achieve a basic thing. Godot should work on TouchInput Components. And stuff releted TouchInput devices.

**Comment by AThousandShips** [comment]
Are you talking about some kind of virtual joystick?

**Reply by llama-nl** [reply]
Yes.

**Comment by Calinou** [comment]
I agree with this, as it's something I noticed while working on https://github.com/godotengine/tps-demo/pull/183. Creating a dual-stick touch control system with good usability is quite a lot of work. This usability is also something that should be as standard as possible across different games made with Godot, as the quality of the touch control implementation can make or break a game.

There are many add-ons that provide such functionality such as https://github.com/archerallstars/G4-Ultimate-Touch-Joypad, but I feel this is integral enough to mobile game development that it has its place in core (almost every first/third-person 3D game will need it).

**Reply by AThousandShips** [reply]
I agree that this is a control that's well within the "can we make this dynamic enough to justify putting it in core" area, some UI elements that are proposed won't work well because they're hard to make dynamic enough to be useful by many, but this is in my opinion not one of those

As long as you can define the action area and configure how the input range works you can make it very customizable

**Comment by syntaxerror247** [comment]
- Consolidating in https://github.com/godotengine/godot-proposals/issues/11192
